tests/test_serialisation.py

```diff
@@ -79,6 +81,61 @@ def test_leaf_serialisation_file(getkey, tmp_path):
     assert tree_loaded_obj is like_obj
 
 
+def test_helpful_errors(getkey, tmp_path):
+    # Test that we get a helpful error message when the loading itself fails
+    tree = jnp.array(1), {}
+    eqx.tree_serialise_leaves(tmp_path, tree)
+    bad_like_tree = jnp.array(2), {
+        "a": jnp.array(2),
+        "b": jnp.array(2),
+    }
+    with pytest.raises(
+        EOFError,
+        match=r"Error at leaf with path \(SequenceKey\(idx=1\), DictKey\(key='a'\)\)",
+    ):
+        _ = eqx.tree_deserialise_leaves(tmp_path, bad_like_tree)
+
+    # Test that we get a helpful error message when the types don't match
+    tree = jnp.array(1), {
+        "a": jnp.array(1),
+        "b": jnp.array(1),
+    }
+    eqx.tree_serialise_leaves(tmp_path, tree)
+    bad_like_tree = jnp.array(2), {
+        "a": jnp.array(2, dtype=jnp.float32),
+        "b": jnp.array(2),
+    }
+    with pytest.raises(
+        RuntimeError,
+        match=r"Deserialised leaf at path \(SequenceKey\(idx=1\), DictKey\(key='a'\)\)",
+    ):
+        _ = eqx.tree_deserialise_leaves(tmp_path, bad_like_tree)
+
+
+def test_generic_dtype_serialisation(getkey, tmp_path):
+    # Ensure we can round trip when we start with an array
+    jax_array = jnp.array(bfloat16(1))
+    eqx.tree_serialise_leaves(tmp_path, jax_array)
+    like_jax_array = jnp.array(bfloat16(2))
+    loaded_jax_array = eqx.tree_deserialise_leaves(tmp_path, like_jax_array)
+    assert eqx.tree_equal(jax_array, loaded_jax_array)
+
+    tree = jnp.array(1), bfloat16(1), np.float32(1), jnp.array(1)
+    like_tree = jnp.array(2), bfloat16(2), np.float32(2), jnp.array(2)
+
+    # Ensure we can round trip when we start with a scalar
+    eqx.tree_serialise_leaves(tmp_path, tree)
+    loaded_tree = eqx.tree_deserialise_leaves(tmp_path, like_tree)
+    # Can't use `tree_equal` because `is_array` treats `bfloat16` as an array and
+    # ends up trying to call `all()` on a `bool`
```

Comment:
What's going on here? This sounds like a potential bug with `tree_equal`?

Reply:
This is related to the https://github.com/patrick-kidger/equinox/issues/507 issue.

```py
        for pytree in pytrees[1:]:
            flat_, treedef_ = jtu.tree_flatten(pytree)
            if treedef_ != treedef:
                return False
            for elem, elem_ in zip(flat, flat_):
                if is_array(elem):
                    if is_array(elem_):
                        if (elem.shape != elem_.shape) or (elem.dtype != elem_.dtype):
                            return False
>                       allsame = (elem == elem_).all()
E                       AttributeError: 'bool' object has no attribute 'all'
```
The `bfloat16`s satisfy the `is_array` check so we end up doing something like `(bfloat16(1) == bfloat16(1)).all()`.

Reply:
Likewise -- `bfloat16` is a JAX-only type, and is treated as an array. That means that `.all()` should exist.

For that matter `np.generic.all` also exists, so again this should work.

How does one obtain a raw Python `bool` here?

Reply:
See the other comment. `jax.dtypes.bfloat16` is different than `jnp.bfloat16`.

Reply:
I think this is indicative of a bug in JAX:
```
x = np.float32(1)
type(x == x)  # np.bool_
y = jax.dtypes.bfloat16(1)
type(y == y)  # Python bool
```
I'd suggest (a) switching `tree_equal` to use `jnp.all` instead of `.all()`; (b) using `tree_equal` here; (c) opening a bug against JAX.

Reply:
Just filed a bug on the `ml_dtypes` library: https://github.com/jax-ml/ml_dtypes/issues/105.